libwnbd/wnbd_ioctl.cpp

```diff
@@ -133,6 +137,315 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
+        index++) {
+
+        if (!SetupDiGetDriverInfoDetailA(AdapterDevHandle, DevInfoData, &DrvInfoData,
+                                         DrvDetailData, DrvDetailData->cbSize, 0)
+            && ERROR_INSUFFICIENT_BUFFER != GetLastError()) {
+            continue;
+        }
+
+        Found = TRUE;
+        break;
+    }
+    if (!Found) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+
+exit:
+    if (ERROR_SUCCESS != Status) {
+        LogError(
+            "Could not get installed driver information data. "
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+    }
+
+    if (Cleanup) {
+        SetupDiDestroyDriverInfoList(AdapterDevHandle, DevInfoData, SPDIT_COMPATDRIVER);
+    }
+
+    return Status;
+}
+
+DWORD WnbdRemoveDriver(HDEVINFO AdapterDevHandle, PSP_DEVINFO_DATA DevInfoData, PBOOL RebootRequired)
+{
+    SP_DRVINFO_DETAIL_DATA_A DrvDetailData = { 0 };
+    DrvDetailData.cbSize = sizeof DrvDetailData;
+    DWORD Status = ERROR_SUCCESS;
+
+    Status = GetDrvDetailData(AdapterDevHandle, DevInfoData, &DrvDetailData);
+    if (ERROR_SUCCESS != Status) {
+        return Status;
+    }
+
+    std::string SearchString(DrvDetailData.InfFileName);
+
+    if (!DiUninstallDevice(0, AdapterDevHandle, DevInfoData, 0, RebootRequired)) {
+        Status = GetLastError();
+    }
+
+    if (SearchString.length()) {
+        SetupUninstallOEMInfA(
+            SearchString.substr(SearchString.find_last_of("\\") + 1).c_str(), SUOI_FORCEDELETE, 0);
+    }
+
+    if (ERROR_SUCCESS != Status) {
+        LogError(
+            "Could not remove driver. "
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+    }
+    return Status;
+}
+
+
+static DWORD FindDevice(HDEVINFO* AdapterDevHandle, SP_DEVINFO_DATA* DevInfoData)
+{
+    CHAR TempBuf[2048];
+    memset(TempBuf, 0, sizeof(TempBuf));
+    BOOL Found = FALSE;
+    DWORD Status = ERROR_FILE_NOT_FOUND;
+
+    *AdapterDevHandle = SetupDiGetClassDevsA(0, 0, 0, DIGCF_ALLCLASSES | DIGCF_PRESENT);
+    if (INVALID_HANDLE_VALUE == *AdapterDevHandle) {
+        Status = GetLastError();
+        LogError(
+            "SetupDiGetClassDevs failed. "
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        return Status;
+    }
+
+    for (DWORD I = 0; !Found && SetupDiEnumDeviceInfo(*AdapterDevHandle, I, DevInfoData); I++)
+    {
+        if (!SetupDiGetDeviceRegistryPropertyA(*AdapterDevHandle, DevInfoData, SPDRP_HARDWAREID, 0,
+            (PBYTE)TempBuf, sizeof(TempBuf) - 1, 0)) {
+            continue;
+        }
+
+        if (strstr(TempBuf, WNBD_HARDWAREID) != NULL) {
+            Found = TRUE;
+            break;
+        }
+    }
+
+    if (!Found) {
+        Status = GetLastError();
+        LogInfo("Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
```

Comment:
This message doesn't provide any context information so it would be very confusing. Let's add "Failed to locate device. Error: ...".